File: front-end/src/domain/entity/workflow.ts
import { UllUUID } from "@ull-tfg/ull-tfg-typescript";
import { WorkFlowName } from "../value-object/workflow-name";
import { Node } from "../value-object/node";
import { NodeName } from "../value-object/node-name";

export class WorkFlow {
  
  private static readonly ERROR_NAME_NOT_DEFINED = "Workflow name not defined";
  private static readonly ERROR_NODES_NOT_DEFINED = "Nodes not defined";
  private static readonly ERROR_NODES_EMPTY = "Nodes cannot be empty";
  private static readonly ERROR_FIRST_NODE_NOT_FOUND = "No initial node found: every node is a destination of at least one link.";
  private static readonly ERROR_MULTIPLE_FIRST_NODES = "Multiple initial nodes found: more than one node has no incoming links.";
  private static readonly ERROR_ID_NOT_DEFINED = "ID not defined";
  private static readonly ERROR_ID_INVALID = "ID must be an instance of UllUUID";
  private static readonly ERROR_CREATION_DATE_INVALID = "Invalid creation date";

  private readonly id: UllUUID;
  private readonly name: WorkFlowName;
  private readonly creationDate: Date;
  private readonly nodes: Node[];
  private readonly firstNode: Node;
  private readonly enabled: boolean;

  constructor(name: WorkFlowName, nodes: Node[], enabled: boolean, creationDate: Date | string = new Date(), id?: UllUUID) {
    this.validateName(name);
    this.validateNodes(nodes);

    if (id) {
      this.validateId(id);
      this.id = id;
    } else {
      this.id = UllUUID.random();
    }

    this.name = name;
    this.nodes = [...nodes];
    this.enabled = enabled;
    this.creationDate = new Date(creationDate);
    this.firstNode = this.computeFirstNode();
  }

  public static from(id: UllUUID, name: WorkFlowName, nodes: Node[], enabled: boolean, creationDate: Date | string = new Date(),
  ): WorkFlow {
    return new WorkFlow(name, nodes, enabled, creationDate, id);
  }

  private validateId(id: UllUUID): void {
    if (!id) throw new Error(WorkFlow.ERROR_ID_NOT_DEFINED);
    if (!(id instanceof UllUUID)) throw new Error(WorkFlow.ERROR_ID_INVALID);
  }

  private validateName(name: WorkFlowName): void {
    if (!name) throw new Error(WorkFlow.ERROR_NAME_NOT_DEFINED);
  }

  private validateNodes(nodes: Node[]): void {
    if (!nodes) throw new Error(WorkFlow.ERROR_NODES_NOT_DEFINED);
    if (nodes.length === 0) throw new Error(WorkFlow.ERROR_NODES_EMPTY);
  }

  private validateCreationDate(creationDate: Date | string): void {
    const date = new Date(creationDate);
    if (isNaN(date.getTime())) {
      throw new Error(WorkFlow.ERROR_CREATION_DATE_INVALID);
    }
  }

  /**
   * @brief compute the first node of the workflow.
   * @returns 
   */
  private computeFirstNode(): Node {
    // Get all node names
    const allNodeNames = new Set(this.nodes.map(node => node.getName()));

    // Get all final nodes from links
    for (const node of this.nodes) {
      for (const link of node.getLinks()) {
        allNodeNames.delete(link.getFinalNode());
      }
    }

    // After removing all final nodes, the remaining names should be the first nodes
    const remainingNames = Array.from(allNodeNames);

    if (remainingNames.length === 0) {
      throw new Error(WorkFlow.ERROR_FIRST_NODE_NOT_FOUND);
    }

    if (remainingNames.length > 1) {
      throw new Error(WorkFlow.ERROR_MULTIPLE_FIRST_NODES);
    }

    const firstNodeName = remainingNames[0];

    const firstNode = this.nodes.find(node => node.getName().equals(firstNodeName));
    if (!firstNode) {
      throw new Error(WorkFlow.ERROR_FIRST_NODE_NOT_FOUND);
    }

    return firstNode;
  }



  getFirstNode(): Node {
    return this.firstNode;
  }

  getId(): UllUUID {
    return this.id;
  }

  getName(): WorkFlowName {
    return this.name;
  }


  getCreationDate(): Date {
    return this.creationDate;
  }

  getNodes(): ReadonlyArray<Node> {
    return this.nodes;
  }


  isEnabled(): boolean {
    return this.enabled;
  }


  setName(name: WorkFlowName): WorkFlow {
    return new WorkFlow(name, this.nodes, this.enabled, this.creationDate, this.id);
  }

  setCreationDate(date: Date): WorkFlow {
    return new WorkFlow(this.name, this.nodes, this.enabled, date, this.id);
  }

  setNodes(nodes: Node[]): WorkFlow {
    return new WorkFlow(this.name, nodes, this.enabled, this.creationDate, this.id);
  }

  activate(): WorkFlow {
    return new WorkFlow(this.name, this.nodes, true, this.creationDate, this.id);
  }

  disactivate(): WorkFlow {
    return new WorkFlow(this.name, this.nodes, false, this.creationDate, this.id);
  }


  toString(): string {
    return `WorkFlow={id=${this.id}, name=${this.name.toString()}, enabled=${this.enabled}, creationDate=${this.creationDate.toISOString()}, nodes=[${this.nodes.map(n => n.toString()).join(', ')}], firstNode=${this.firstNode.toString()}}`;
  }
}

File: back-end/src/main/java/es/ull/project/domain/entity/WorkFlow.java
package es.ull.project.domain.entity;

import es.ull.project.domain.valueobject.WorkFlowName;
import es.ull.project.domain.valueobject.Link;
import es.ull.project.domain.valueobject.Node;
import es.ull.project.domain.valueobject.NodeName;

import java.util.UUID;
import java.util.stream.Collectors;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * @brief Represents a workflow with a list of nodes.
 */
public class WorkFlow {

    private static final String ERROR_NAME_NOT_DEFINED = "Workflow name not defined";
    private static final String ERROR_NODES_NOT_DEFINED = "Nodes not defined";
    private static final String ERROR_NODES_EMPTY = "Nodes cannot be empty";
    private static final String ERROR_DATE_NOT_DEFINED = "Date not defined";
    private static final String ERROR_FIRST_NODE_NOT_FOUND = "No initial node found: every node is a destination of at least one link.";
    private static final String ERROR_MULTIPLE_FIRST_NODES = "Multiple initial nodes found: more than one node has no incoming links.";

    private final UUID id;
    private final WorkFlowName name;
    private final LocalDateTime creationDate;
    private final List<Node> nodes;
    private final boolean enabled;
    /**
     * computed atribute
     */
    private Node firstNode;

    public WorkFlow(
            WorkFlowName name,
            List<Node> nodes, boolean enabled) {
        validateName(name);
        validateNodes(nodes);

        this.id = UUID.randomUUID();
        this.name = name;
        this.nodes = new ArrayList<>(nodes);
        this.creationDate = LocalDateTime.now();
        this.firstNode = computeFirstNode();
        this.enabled = enabled;
    }

    public WorkFlow(
            UUID id,
            WorkFlowName name,
            List<Node> nodes,
            LocalDateTime creationDate, boolean enabled) {

        validateName(name);
        validateNodes(nodes);
        validateDate(creationDate);
        this.id = id;
        this.name = name;
        this.nodes = new ArrayList<>(nodes);
        this.creationDate = creationDate;
        this.firstNode = computeFirstNode();
        this.enabled = enabled;
    }

    private void validateName(WorkFlowName name) {
        if (name == null) {
            throw new IllegalArgumentException(ERROR_NAME_NOT_DEFINED);
        }
    }

    private void validateNodes(List<Node> nodes) {
        if (nodes == null) {
            throw new IllegalArgumentException(ERROR_NODES_NOT_DEFINED);
        }
        if (nodes.isEmpty()) {
            throw new IllegalArgumentException(ERROR_NODES_EMPTY);
        }
    }

    private void validateDate(LocalDateTime date) {
        if (date == null) {
            throw new IllegalArgumentException(ERROR_DATE_NOT_DEFINED);
        }
    }

    /**
     * @brief computes the first node of the workflow.
     * @return
     */
    private Node computeFirstNode() {
        Set<NodeName> allNodes = nodes.stream()
                .map(Node::getName)
                .collect(Collectors.toSet());
        Set<NodeName> destinationNodes = new HashSet<>();
        for (Node node : nodes) {
            for (Link link : node.getLinks()) {
                destinationNodes.add(link.getFinalNode());
            }
        }
        // Remove all destination nodes from the set of all nodes
        allNodes.removeAll(destinationNodes);
        if (allNodes.isEmpty()) {
            throw new IllegalStateException(ERROR_FIRST_NODE_NOT_FOUND);
        } else if (allNodes.size() > 1) {
            throw new IllegalStateException(ERROR_MULTIPLE_FIRST_NODES);
        }
        NodeName firstNodeName = allNodes.iterator().next();
        return nodes.stream()
                .filter(node -> node.getName().equals(firstNodeName))
                .findFirst()
                .orElseThrow(() -> new IllegalStateException(ERROR_FIRST_NODE_NOT_FOUND));
    }

    public Node getFirstNode() {
        return this.firstNode;
    }

    public UUID getId() {
        return this.id;
    }

    public WorkFlowName getName() {
        return this.name;
    }

    public LocalDateTime getCreationDate() {
        return this.creationDate;
    }

    public List<Node> getNodes() {
        return Collections.unmodifiableList(this.nodes);
    }

    public boolean isEnabled() {
        return this.enabled;
    }

    public WorkFlow setName(WorkFlowName name) {
        return new WorkFlow(this.id, name, this.nodes, this.creationDate, this.enabled);
    }

    public WorkFlow setCreationDate(LocalDateTime creationDate) {
        return new WorkFlow(this.id, this.name, this.nodes, creationDate, this.enabled);
    }

    public WorkFlow setNodes(List<Node> nodes) {
        return new WorkFlow(this.id, this.name, nodes, this.creationDate, this.enabled);
    }

    public WorkFlow setEnabled(boolean enabled) {
        return new WorkFlow(this.id, this.name, this.nodes, this.creationDate, enabled);
    }

    public WorkFlow activate() {
        return new WorkFlow(this.id, this.name, this.nodes, this.creationDate, true);
    }

    public WorkFlow disactivate() {
        return new WorkFlow(this.id, this.name, this.nodes, this.creationDate, false);
    }

    @Override
    public String toString() {
        return String.format(
                "Workflow={id=%s, name=%s, creationDate=%s, enabled=%s, nodes=%s, firstNode=%s}",
                this.id, this.name, this.creationDate, this.enabled, this.nodes, this.firstNode);
    }
}
